File: Emulator/src/main/core/net/EventProducer.java
package core.net;

import java.nio.ByteBuffer;

/**
 * The interface Event producer.
 */
public interface EventProducer {

    /**
     * Produce reader io read event.
     *
     * @param session the session
     * @param buffer  the buffer
     * @return the io read event
     */
    IoReadEvent produceReader(IoSession session, ByteBuffer buffer);

    /**
     * Produce writer io write event.
     *
     * @param session the session
     * @param context the context
     * @return the io write event
     */
    IoWriteEvent produceWriter(IoSession session, Object context);

}

File: game-server/src/main/core/net/EventProducer.kt
package core.net

import java.nio.ByteBuffer

/**
 * Used for producing I/O events.
 *
 * @author Emperor
 */
interface EventProducer {

    /**
     * Produces a new read event.
     *
     * @param session The session.
     * @param buffer  The buffer to read.
     * @return The read event handler.
     */
    fun produceReader(session: IoSession, buffer: ByteBuffer): IoReadEvent

    /**
     * Produces a new writing event.
     *
     * @param session The session.
     * @param context The context.
     * @return The write event handler.
     */
    fun produceWriter(session: IoSession, context: Any): IoWriteEvent
}
